File: apps/api-rs/src/rag/vector_pipeline.rs
use std::collections::HashSet;
use std::time::Duration;

use anyhow::{anyhow, bail, Context, Result};
use chrono::Utc;
use serde::Serialize;
use serde_json::json;
use sqlx::{PgPool, Row};
use tracing::{error, info, warn};
use uuid::Uuid;

use crate::config::EmbeddingConfig;
use crate::rag::embedding::{EmbeddingClient, EmbeddingClientConfig};
use crate::rag::vector_document::index_document;
use crate::rag::vector_index::{
    physical_index_name, ElasticsearchChunkIndexer, ElasticsearchConfig,
};
use crate::rag::{vector_jobs, vector_store};

#[derive(Debug, Clone, Serialize)]
pub struct VectorConsistency {
    pub index_alias: String,
    pub physical_index: Option<String>,
    pub expected_chunks: i64,
    pub actual_chunks: i64,
    pub missing_chunks: i64,
    pub stale_chunks: i64,
    pub missing_or_stale_chunks: i64,
    pub consistent: bool,
}

pub fn start_vector_worker(
    pool: PgPool,
    config: EmbeddingConfig,
    es_url: Option<String>,
    rabbitmq_url: Option<String>,
) {
    if !config.enabled {
        info!("vector worker disabled because embedding is disabled");
        return;
    }
    let Some(es_url) = es_url else {
        error!("vector worker disabled because ELASTICSEARCH_URL is missing");
        return;
    };
    tokio::spawn(async move {
        if let Err(error) = run_worker(pool, config, es_url, rabbitmq_url).await {
            error!(error = %error, "vector worker stopped unexpectedly");
        }
    });
}

pub async fn enqueue_document(
    pool: &PgPool,
    tenant_id: Uuid,
    kb_id: Uuid,
    doc_id: Uuid,
    parse_job_id: Uuid,
    config: &EmbeddingConfig,
    force: bool,
) -> Result<Uuid> {
    let target = vector_jobs::active_index(pool, &config.index_alias)
        .await?
        .unwrap_or_else(|| desired_index(config));
    vector_jobs::enqueue_document(
        pool,
        tenant_id,
        kb_id,
        doc_id,
        parse_job_id,
        &target,
        config,
        force,
    )
    .await
}

pub async fn schedule_rebuild(pool: &PgPool, config: &EmbeddingConfig) -> Result<(Uuid, String)> {
    if has_open_rebuild(pool, &config.index_alias).await? {
        let row = sqlx::query(
            "SELECT id, target_index FROM vector_jobs
             WHERE operation = 'rebuild_index' AND status IN ('pending', 'running')
               AND embedding_model = $1
             ORDER BY created_at DESC LIMIT 1",
        )
        .bind(&config.model)
        .fetch_one(pool)
        .await?;
        return Ok((row.try_get("id")?, row.try_get("target_index")?));
    }
    let base = desired_index(config);
    let active = vector_jobs::active_index(pool, &config.index_alias).await?;
    let target = if active.as_deref() == Some(base.as_str()) {
        format!("{base}-r{}", Utc::now().timestamp_millis())
    } else {
        base
    };
    let mut tx = pool.begin().await?;
    vector_jobs::create_building_version(&mut tx, &config.index_alias, &target, config).await?;
    tx.commit().await?;
    let id = vector_jobs::enqueue_rebuild(pool, &target, config).await?;
    Ok((id, target))
}

pub async fn consistency(
    pool: &PgPool,
    config: &EmbeddingConfig,
    es_url: &str,
) -> Result<VectorConsistency> {
    let physical_index = vector_jobs::active_index(pool, &config.index_alias).await?;
    let expected_ids = expected_chunk_ids(pool).await?;
    let actual_ids = if let Some(index) = &physical_index {
        indexer(es_url, index, config)?.chunk_ids().await?
    } else {
        HashSet::new()
    };
    let missing = expected_ids.difference(&actual_ids).count() as i64;
    let stale = actual_ids.difference(&expected_ids).count() as i64;
    Ok(VectorConsistency {
        index_alias: config.index_alias.clone(),
        physical_index,
        expected_chunks: expected_ids.len() as i64,
        actual_chunks: actual_ids.len() as i64,
        missing_chunks: missing,
        stale_chunks: stale,
        missing_or_stale_chunks: missing + stale,
        consistent: missing == 0 && stale == 0,
    })
}

pub async fn quick_consistency(
    pool: &PgPool,
    config: &EmbeddingConfig,
    es_url: &str,
) -> Result<VectorConsistency> {
    let physical_index = vector_jobs::active_index(pool, &config.index_alias).await?;
    let expected: i64 = sqlx::query_scalar(
        "SELECT COUNT(*)::bigint
         FROM chunks c
         JOIN documents d ON d.id = c.doc_id AND d.latest_parse_job_id = c.parse_job_id
         WHERE d.parse_status = 'indexed'",
    )
    .fetch_one(pool)
    .await?;
    let actual = if let Some(index) = &physical_index {
        indexer(es_url, index, config)?.count().await? as i64
    } else {
        0
    };
    let missing = (expected - actual).max(0);
    let stale = (actual - expected).max(0);
    Ok(VectorConsistency {
        index_alias: config.index_alias.clone(),
        physical_index,
        expected_chunks: expected,
        actual_chunks: actual,
        missing_chunks: missing,
        stale_chunks: stale,
        missing_or_stale_chunks: missing + stale,
        consistent: expected == actual,
    })
}

async fn run_worker(
    pool: PgPool,
    config: EmbeddingConfig,
    es_url: String,
    rabbitmq_url: Option<String>,
) -> Result<()> {
    let client_config = EmbeddingClientConfig::try_from(&config)?;
    let embedding_client = EmbeddingClient::new(client_config)?;
    let worker_id = format!("vector-worker-{}", Uuid::new_v4());
    vector_store::reconcile_legacy_embeddings(&pool).await?;
    recover_expired_leases(&pool).await?;
    bootstrap_jobs(&pool, &config, &es_url).await?;
    let (queue_sender, mut queue_receiver) = tokio::sync::mpsc::channel(1_024);
    if let Some(rabbitmq_url) = rabbitmq_url {
        crate::rag::vector_queue::start(rabbitmq_url, pool.clone(), queue_sender);
    }

    let poll = Duration::from_millis(config.worker_poll_ms.clamp(250, 60_000));
    let lease_recovery_ticks = (30_000 / poll.as_millis().max(1) as u64).max(1);
    let mut idle_ticks = 0_u64;
    loop {
        let queued_job_id = queue_receiver.try_recv().ok();
        let claimed = if let Some(job_id) = queued_job_id {
            vector_jobs::claim_by_id(&pool, &worker_id, job_id).await?
        } else {
            vector_jobs::claim_next(&pool, &worker_id).await?
        };
        match claimed {
            Some(job) => {
                idle_ticks = 0;
                let result =
                    process_job(&pool, &config, &es_url, &embedding_client, &worker_id, &job).await;
                match result {
                    Ok(metadata) => {
                        vector_jobs::complete(&pool, job.id, metadata).await?;
                    }
                    Err(error) => {
                        let message = format!("{error:#}");
                        let retry = vector_jobs::fail(&pool, &job, &message).await?;
                        warn!(job_id = %job.id, operation = %job.operation, retry, error = %message, "vector job failed");
                        if !retry {
                            if let Some(doc_id) = job.doc_id {
                                vector_store::mark_document_terminal_failure(
                                    &pool,
                                    doc_id,
                                    job.parse_job_id,
                                    &config.model,
                                    &message,
                                )
                                .await?;
                            }
                            if job.operation == "rebuild_index" {
                                vector_jobs::mark_version_failed(
                                    &pool,
                                    &job.target_index,
                                    &message,
                                )
                                .await?;
                            }
                        }
                    }
                }
            }
            None => {
                idle_ticks += 1;
                if idle_ticks % lease_recovery_ticks == 0 {
                    recover_expired_leases(&pool).await?;
                }
                if idle_ticks.saturating_mul(poll.as_millis() as u64) >= 60_000 {
                    idle_ticks = 0;
                    if let Err(error) = ensure_consistency(&pool, &config, &es_url).await {
                        warn!(error = %error, "periodic vector consistency check failed");
                    }
                }
                tokio::time::sleep(poll).await;
            }
        }
    }
}

async fn recover_expired_leases(pool: &PgPool) -> Result<()> {
    let recovered = vector_jobs::recover_leases(pool).await?;
    if recovered > 0 {
        warn!(recovered, "recovered expired vector job leases");
    }
    Ok(())
}

async fn bootstrap_jobs(pool: &PgPool, config: &EmbeddingConfig, es_url: &str) -> Result<()> {
    let desired = desired_index(config);
    let desired_indexer = indexer(es_url, &desired, config)?;
    let alias_targets = desired_indexer.alias_targets().await?;
    let active = vector_jobs::active_index(pool, &config.index_alias).await?;
    let alias_ready = alias_targets.len() == 1 && alias_targets.first() == Some(&desired);
    if active.as_deref() != Some(desired.as_str()) || !alias_ready {
        schedule_rebuild(pool, config).await?;
    } else {
        ensure_consistency(pool, config, es_url).await?;
    }

    let rows = sqlx::query(
        "SELECT tenant_id, kb_id, id, latest_parse_job_id
         FROM documents
         WHERE parse_status IN ('chunked', 'embedding')
           AND latest_parse_job_id IS NOT NULL AND chunk_count > 0",
    )
    .fetch_all(pool)
    .await?;
    for row in rows {
        enqueue_document(
            pool,
            row.try_get("tenant_id")?,
            row.try_get("kb_id")?,
            row.try_get("id")?,
            row.try_get("latest_parse_job_id")?,
            config,
            true,
        )
        .await?;
    }
    Ok(())
}

async fn process_job(
    pool: &PgPool,
    config: &EmbeddingConfig,
    es_url: &str,
    embedding_client: &EmbeddingClient,
    worker_id: &str,
    job: &vector_jobs::VectorJob,
) -> Result<serde_json::Value> {
    if job.embedding_model != config.model || job.embedding_dim != config.dimension {
        bail!("vector job model or dimension no longer matches runtime configuration");
    }
    match job.operation.as_str() {
        "index_document" => {
            let doc_id = job.doc_id.context("index_document job is missing doc_id")?;
            let parse_job_id = job
                .parse_job_id
                .context("index_document job is missing parse_job_id")?;
            let target = vector_jobs::active_index(pool, &config.index_alias)
                .await?
                .unwrap_or_else(|| job.target_index.clone());
            let outcome = index_document(
                pool,
                embedding_client,
                &indexer(es_url, &target, config)?,
                doc_id,
                parse_job_id,
                config,
            )
            .await?;
            Ok(json!({
                "physical_index": target,
                "indexed_chunks": outcome.indexed_chunks,
                "generated_embeddings": outcome.generated_embeddings,
                "reused_embeddings": outcome.reused_embeddings,
                "skipped": outcome.skipped,
            }))
        }
        "rebuild_index" => {
            rebuild_index(pool, config, es_url, embedding_client, worker_id, job).await
        }
        operation => Err(anyhow!("unsupported vector job operation {operation}")),
    }
}

async fn rebuild_index(
    pool: &PgPool,
    config: &EmbeddingConfig,
    es_url: &str,
    embedding_client: &EmbeddingClient,
    worker_id: &str,
    job: &vector_jobs::VectorJob,
) -> Result<serde_json::Value> {
    let indexer = indexer(es_url, &job.target_index, config)?;
    let attached = indexer.alias_targets().await?;
    if attached.iter().any(|index| index == &job.target_index) {
        let (expected, actual) = verify_index_contents(pool, &indexer).await?;
        let previous = indexer.switch_alias().await?;
        vector_jobs::activate_version(
            pool,
            &config.index_alias,
            &job.target_index,
            expected,
            actual,
        )
        .await?;
        for retired in previous.iter().filter(|index| *index != &job.target_index) {
            indexer.delete_index(retired).await?;
        }
        for retired in vector_jobs::retired_indexes(pool, &config.index_alias).await? {
            if retired != job.target_index {
                indexer.delete_index(&retired).await?;
            }
        }
        if config.index_name != job.target_index && !previous.contains(&config.index_name) {
            indexer.delete_index(&config.index_name).await?;
        }
        return Ok(json!({
            "physical_index": job.target_index,
            "expected_chunks": expected,
            "actual_chunks": actual,
            "recovered_alias_activation": true,
            "retired_indices": previous,
        }));
    }
    indexer.reset_inactive_index(config.dimension).await?;
    let documents = sqlx::query(
        "SELECT id, latest_parse_job_id
         FROM documents
         WHERE parse_status = 'indexed'
           AND latest_parse_job_id IS NOT NULL AND chunk_count > 0
         ORDER BY updated_at, id",
    )
    .fetch_all(pool)
    .await?;
    let mut indexed_documents = 0usize;
    for row in documents {
        vector_jobs::heartbeat(pool, job.id, worker_id).await?;
        let outcome = index_document(
            pool,
            embedding_client,
            &indexer,
            row.try_get("id")?,
            row.try_get("latest_parse_job_id")?,
            config,
        )
        .await?;
        if !outcome.skipped {
            indexed_documents += 1;
        }
    }
    indexer.refresh().await?;
    let (expected, actual) = verify_index_contents(pool, &indexer).await?;
    let previous = indexer.switch_alias().await?;
    vector_jobs::activate_version(
        pool,
        &config.index_alias,
        &job.target_index,
        expected,
        actual,
    )
    .await?;
    for retired in previous.iter().filter(|index| *index != &job.target_index) {
        indexer.delete_index(retired).await?;
    }
    info!(physical_index = %job.target_index, expected, actual, "activated rebuilt vector index");
    Ok(json!({
        "physical_index": job.target_index,
        "indexed_documents": indexed_documents,
        "expected_chunks": expected,
        "actual_chunks": actual,
        "retired_indices": previous,
    }))
}

async fn verify_index_contents(
    pool: &PgPool,
    indexer: &ElasticsearchChunkIndexer,
) -> Result<(i64, i64)> {
    let expected_ids = expected_chunk_ids(pool).await?;
    let actual_ids = indexer.chunk_ids().await?;
    let missing = expected_ids.difference(&actual_ids).count();
    let stale = actual_ids.difference(&expected_ids).count();
    if missing > 0 || stale > 0 {
        bail!("vector index has {missing} missing and {stale} stale chunks");
    }
    Ok((expected_ids.len() as i64, actual_ids.len() as i64))
}

async fn ensure_consistency(pool: &PgPool, config: &EmbeddingConfig, es_url: &str) -> Result<()> {
    let snapshot = consistency(pool, config, es_url).await?;
    if snapshot.consistent {
        if let Some(index) = snapshot.physical_index.as_deref() {
            vector_store::mark_current_embeddings_indexed(pool, &config.model, index).await?;
            vector_jobs::refresh_active_version_counts(pool, index, snapshot.actual_chunks).await?;
        }
    } else if !has_open_rebuild(pool, &config.index_alias).await? {
        warn!(
            expected = snapshot.expected_chunks,
            actual = snapshot.actual_chunks,
            "vector index drift detected; scheduling rebuild"
        );
        schedule_rebuild(pool, config).await?;
    }
    Ok(())
}

async fn expected_chunk_ids(pool: &PgPool) -> Result<HashSet<Uuid>> {
    let ids: Vec<Uuid> = sqlx::query_scalar(
        "SELECT c.id
         FROM chunks c
         JOIN documents d ON d.id = c.doc_id AND d.latest_parse_job_id = c.parse_job_id
         WHERE d.parse_status = 'indexed'",
    )
    .fetch_all(pool)
    .await
    .context("failed to load current indexed chunk identifiers")?;
    Ok(ids.into_iter().collect())
}

async fn has_open_rebuild(pool: &PgPool, alias: &str) -> Result<bool> {
    let exists: bool = sqlx::query_scalar(
        "SELECT EXISTS(
            SELECT 1 FROM vector_jobs j
            JOIN vector_index_versions v ON v.physical_index = j.target_index
            WHERE j.operation = 'rebuild_index' AND j.status IN ('pending', 'running')
              AND v.index_alias = $1
         )",
    )
    .bind(alias)
    .fetch_one(pool)
    .await?;
    Ok(exists)
}

fn desired_index(config: &EmbeddingConfig) -> String {
    let version = config.index_schema_version;
    physical_index_name(&config.index_name, &config.model, config.dimension, version)
}

fn indexer(
    es_url: &str,
    physical_index: &str,
    config: &EmbeddingConfig,
) -> Result<ElasticsearchChunkIndexer> {
    ElasticsearchChunkIndexer::new(ElasticsearchConfig {
        base_url: es_url.to_string(),
        index_name: physical_index.to_string(),
        alias_name: config.index_alias.clone(),
        timeout_seconds: 120,
    })
}

File: apps/api-bun/src/rag/vector_pipeline.ts
// 移植自 apps/api-rs/src/rag/vector_pipeline.rs
// —— VectorConsistency 快照 / 物理索引命名与 alias 切换 / 重建调度 / 后台 worker 主循环。
//
// 与 Rust 的差异（明确记录，不做静默兜底）：
// Rust 用 vector_queue.rs（lapin / RabbitMQ）作为可选唤醒通道，DB 轮询始终在跑；
// Bun 端口未引入 AMQP 客户端依赖，startVectorWorker 的 rabbitmqUrl 参数只记录告警，
// 作业消费依赖 claimNext 数据库轮询 —— 等价于 Rust 在 "RabbitMQ 不可用，database polling
// remains active" 分支的行为。
import type { Sql } from 'postgres';
import type { EmbeddingConfig } from '../config.ts';
import { newUuid } from '../infra/uuid.ts';
import { EmbeddingClient, embeddingClientConfigFrom } from './embedding.ts';
import { ElasticsearchChunkIndexer } from './vector_index.ts';
import { physicalIndexName } from './vector_index/schema.ts';
import * as vectorJobs from './vector_jobs.ts';
import { processJob, setDifference } from './vector_pipeline/jobs.ts';
import * as vectorStore from './vector_store.ts';

/** 与 Rust VectorConsistency 的 serde 输出逐字段对齐（snake_case）。 */
export interface VectorConsistency {
  index_alias: string;
  physical_index: string | null;
  expected_chunks: number;
  actual_chunks: number;
  missing_chunks: number;
  stale_chunks: number;
  missing_or_stale_chunks: number;
  consistent: boolean;
}

const WORKER_LEASE_RECOVERY_MS = 30_000;
const WORKER_CONSISTENCY_INTERVAL_MS = 60_000;

/**
 * 对应 Rust start_vector_worker：校验前置条件后 fire-and-forget 启动后台循环。
 * 与 Rust 相同，worker 异常退出只记录日志，由入口决定是否重启。
 */
export function startVectorWorker(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string | null,
  rabbitmqUrl: string | null,
): void {
  if (!config.enabled) {
    console.log('[documind][rag] vector worker disabled because embedding is disabled');
    return;
  }
  if (esUrl === null) {
    console.error('[documind][rag] vector worker disabled because ELASTICSEARCH_URL is missing');
    return;
  }
  if (rabbitmqUrl !== null) {
    console.warn(
      '[documind][rag] RabbitMQ vector queue is not available in the Bun port; database polling remains active',
    );
  }
  void runWorker(sql, config, esUrl).catch((error: unknown) => {
    console.error(
      `[documind][rag] vector worker stopped unexpectedly: ${describeError(error)}`,
    );
  });
}

/** 对应 Rust enqueue_document：目标索引取当前 active 索引，缺失时用期望索引。 */
export async function enqueueDocument(
  sql: Sql,
  tenantId: string,
  kbId: string,
  docId: string,
  parseJobId: string,
  config: EmbeddingConfig,
  force: boolean,
): Promise<string> {
  const target = (await vectorJobs.activeIndex(sql, config.indexAlias)) ?? desiredIndex(config);
  return vectorJobs.enqueueDocument(
    sql, tenantId, kbId, docId, parseJobId, target, config, force,
  );
}

/** 对应 Rust schedule_rebuild：返回 [job_id, target_index]。 */
export async function scheduleRebuild(
  sql: Sql,
  config: EmbeddingConfig,
): Promise<[string, string]> {
  if (await hasOpenRebuild(sql, config.indexAlias)) {
    const rows = await sql.unsafe(
      `SELECT id, target_index FROM vector_jobs
       WHERE operation = 'rebuild_index' AND status IN ('pending', 'running')
         AND embedding_model = $1
       ORDER BY created_at DESC LIMIT 1`,
      [config.model],
    );
    const row = rows[0] as Record<string, unknown> | undefined;
    if (row === undefined) {
      throw new Error('open rebuild job disappeared while scheduling a rebuild');
    }
    return [asColumnString(row.id, 'id'), asColumnString(row.target_index, 'target_index')];
  }
  const base = desiredIndex(config);
  const active = await vectorJobs.activeIndex(sql, config.indexAlias);
  const target = active === base ? `${base}-r${Date.now()}` : base;
  await sql.begin(async (tx) => {
    await vectorJobs.createBuildingVersion(tx, config.indexAlias, target, config);
  });
  const id = await vectorJobs.enqueueRebuild(sql, target, config);
  return [id, target];
}

/** 对应 Rust consistency：逐 chunk 比对（可发现 stale），需要 ES scroll。 */
export async function consistency(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string,
): Promise<VectorConsistency> {
  const physicalIndex = await vectorJobs.activeIndex(sql, config.indexAlias);
  const expectedIds = await expectedChunkIds(sql);
  const actualIds = physicalIndex === null
    ? new Set<string>()
    : await indexer(esUrl, physicalIndex, config).chunkIds();
  const missing = setDifference(expectedIds, actualIds).size;
  const stale = setDifference(actualIds, expectedIds).size;
  return {
    index_alias: config.indexAlias,
    physical_index: physicalIndex,
    expected_chunks: expectedIds.size,
    actual_chunks: actualIds.size,
    missing_chunks: missing,
    stale_chunks: stale,
    missing_or_stale_chunks: missing + stale,
    consistent: missing === 0 && stale === 0,
  };
}

/** 对应 Rust quick_consistency：只比数量，健康检查用。 */
export async function quickConsistency(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string,
): Promise<VectorConsistency> {
  const physicalIndex = await vectorJobs.activeIndex(sql, config.indexAlias);
  const rows = await sql.unsafe(
    `SELECT COUNT(*)::bigint AS count
     FROM chunks c
     JOIN documents d ON d.id = c.doc_id AND d.latest_parse_job_id = c.parse_job_id
     WHERE d.parse_status = 'indexed'`,
  );
  const row = rows[0] as Record<string, unknown> | undefined;
  if (row === undefined) throw new Error('expected chunk count query returned no row');
  const expected = Number(row.count ?? 0);
  const actual = physicalIndex === null
    ? 0
    : await indexer(esUrl, physicalIndex, config).count();
  const missing = Math.max(0, expected - actual);
  const stale = Math.max(0, actual - expected);
  return {
    index_alias: config.indexAlias,
    physical_index: physicalIndex,
    expected_chunks: expected,
    actual_chunks: actual,
    missing_chunks: missing,
    stale_chunks: stale,
    missing_or_stale_chunks: missing + stale,
    consistent: expected === actual,
  };
}

/** 对应 Rust expected_chunk_ids（jobs.ts 的 verify_index_contents 也复用）。 */
export async function expectedChunkIds(sql: Sql): Promise<Set<string>> {
  try {
    const rows = await sql.unsafe(
      `SELECT c.id
       FROM chunks c
       JOIN documents d ON d.id = c.doc_id AND d.latest_parse_job_id = c.parse_job_id
       WHERE d.parse_status = 'indexed'`,
    );
    const ids = new Set<string>();
    for (const raw of rows) {
      ids.add(asColumnString((raw as Record<string, unknown>).id, 'id'));
    }
    return ids;
  } catch (error) {
    throw new Error(
      `failed to load current indexed chunk identifiers: ${describeError(error)}`,
    );
  }
}

/** 对应 Rust indexer：每个物理索引一个 indexer，alias 始终指向 config.index_alias。 */
export function indexer(
  esUrl: string,
  physicalIndex: string,
  config: EmbeddingConfig,
): ElasticsearchChunkIndexer {
  return new ElasticsearchChunkIndexer({
    baseUrl: esUrl,
    indexName: physicalIndex,
    aliasName: config.indexAlias,
    timeoutSeconds: 120,
  });
}

/** 对应 Rust desired_index。 */
export function desiredIndex(config: EmbeddingConfig): string {
  return physicalIndexName(
    config.indexName,
    config.model,
    config.dimension,
    config.indexSchemaVersion,
  );
}

/** 对应 Rust run_worker：claim → process → complete/fail，空转时周期性 lease 回收与一致性校验。 */
async function runWorker(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string,
): Promise<void> {
  const embeddingClient = new EmbeddingClient(embeddingClientConfigFrom(config));
  const workerId = `vector-worker-${newUuid()}`;
  await vectorStore.reconcileLegacyEmbeddings(sql);
  await recoverExpiredLeases(sql);
  await bootstrapJobs(sql, config, esUrl);

  const poll = Math.min(60_000, Math.max(250, config.workerPollMs));
  const leaseRecoveryTicks = Math.max(1, Math.floor(WORKER_LEASE_RECOVERY_MS / Math.max(1, poll)));
  let idleTicks = 0;
  for (;;) {
    const job = await vectorJobs.claimNext(sql, workerId);
    if (job === null) {
      idleTicks += 1;
      if (idleTicks % leaseRecoveryTicks === 0) {
        await recoverExpiredLeases(sql);
      }
      if (idleTicks * poll >= WORKER_CONSISTENCY_INTERVAL_MS) {
        idleTicks = 0;
        try {
          await ensureConsistency(sql, config, esUrl);
        } catch (error) {
          console.warn(
            `[documind][rag] periodic vector consistency check failed: ${describeError(error)}`,
          );
        }
      }
      await sleep(poll);
      continue;
    }
    idleTicks = 0;
    const result = await processJob(sql, config, esUrl, embeddingClient, workerId, job).then(
      (metadata) => ({ ok: true as const, metadata }),
      (error: unknown) => ({ ok: false as const, error }),
    );
    if (result.ok) {
      await vectorJobs.complete(sql, job.id, result.metadata);
      continue;
    }
    const message = describeError(result.error);
    const retry = await vectorJobs.fail(sql, job, message);
    console.warn(
      `[documind][rag] vector job failed job_id=${job.id} operation=${job.operation} retry=${retry} error=${message}`,
    );
    if (!retry) {
      if (job.docId !== null) {
        await vectorStore.markDocumentTerminalFailure(
          sql, job.docId, job.parseJobId, config.model, message,
        );
      }
      if (job.operation === 'rebuild_index') {
        await vectorJobs.markVersionFailed(sql, job.targetIndex, message);
      }
    }
  }
}

async function recoverExpiredLeases(sql: Sql): Promise<void> {
  const recovered = await vectorJobs.recoverLeases(sql);
  if (recovered > 0) {
    console.warn(`[documind][rag] recovered expired vector job leases: ${recovered}`);
  }
}

/** 对应 Rust bootstrap_jobs：先保证 alias 指向期望索引，再补排队 chunked/embedding 文档。 */
async function bootstrapJobs(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string,
): Promise<void> {
  const desired = desiredIndex(config);
  const aliasTargets = await indexer(esUrl, desired, config).aliasTargets();
  const active = await vectorJobs.activeIndex(sql, config.indexAlias);
  const aliasReady = aliasTargets.length === 1 && aliasTargets[0] === desired;
  if (active !== desired || !aliasReady) {
    await scheduleRebuild(sql, config);
  } else {
    await ensureConsistency(sql, config, esUrl);
  }

  const rows = await sql.unsafe(
    `SELECT tenant_id, kb_id, id, latest_parse_job_id
     FROM documents
     WHERE parse_status IN ('chunked', 'embedding')
       AND latest_parse_job_id IS NOT NULL AND chunk_count > 0`,
  );
  for (const raw of rows) {
    const row = raw as Record<string, unknown>;
    await enqueueDocument(
      sql,
      asColumnString(row.tenant_id, 'tenant_id'),
      asColumnString(row.kb_id, 'kb_id'),
      asColumnString(row.id, 'id'),
      asColumnString(row.latest_parse_job_id, 'latest_parse_job_id'),
      config,
      true,
    );
  }
}

/** 对应 Rust ensure_consistency：一致则回填 embedding index 状态，漂移则调度重建。 */
async function ensureConsistency(
  sql: Sql,
  config: EmbeddingConfig,
  esUrl: string,
): Promise<void> {
  const snapshot = await consistency(sql, config, esUrl);
  if (snapshot.consistent) {
    if (snapshot.physical_index !== null) {
      await vectorStore.markCurrentEmbeddingsIndexed(sql, config.model, snapshot.physical_index);
      await vectorJobs.refreshActiveVersionCounts(
        sql, snapshot.physical_index, snapshot.actual_chunks,
      );
    }
  } else if (!(await hasOpenRebuild(sql, config.indexAlias))) {
    console.warn(
      `[documind][rag] vector index drift detected; scheduling rebuild expected=${snapshot.expected_chunks} actual=${snapshot.actual_chunks}`,
    );
    await scheduleRebuild(sql, config);
  }
}

async function hasOpenRebuild(sql: Sql, alias: string): Promise<boolean> {
  const rows = await sql.unsafe(
    `SELECT EXISTS(
        SELECT 1 FROM vector_jobs j
        JOIN vector_index_versions v ON v.physical_index = j.target_index
        WHERE j.operation = 'rebuild_index' AND j.status IN ('pending', 'running')
          AND v.index_alias = $1
     ) AS exists`,
    [alias],
  );
  const row = rows[0] as Record<string, unknown> | undefined;
  if (row === undefined) throw new Error('open rebuild check returned no row');
  return row.exists === true;
}

function asColumnString(value: unknown, column: string): string {
  if (typeof value !== 'string') {
    throw new Error(`unexpected non-string column ${column} in vector pipeline row`);
  }
  return value;
}

function sleep(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

/** 与 Rust anyhow "{error:#}" 对齐：拼接 Error.cause 链。 */
function describeError(error: unknown): string {
  if (error instanceof Error) {
    const messages: string[] = [error.message];
    let cause: unknown = error.cause;
    while (cause instanceof Error) {
      messages.push(cause.message);
      cause = cause.cause;
    }
    return messages.join(': ');
  }
  return String(error);
}
